common/sgx/quote.c

```diff
@@ -302,3 +376,311 @@ oe_result_t oe_verify_quote_internal(
     oe_cert_chain_free(&pck_cert_chain);
     return result;
 }
+
+oe_result_t oe_get_quote_cert_chain_internal(
+    const uint8_t* quote,
+    const size_t quote_size,
+    const uint8_t** pem_pck_certificate,
+    size_t* pem_pck_certificate_size,
+    oe_cert_chain_t* pck_cert_chain)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    sgx_quote_t* sgx_quote = NULL;
+    sgx_quote_auth_data_t* quote_auth_data = NULL;
+    sgx_qe_auth_data_t qe_auth_data = {0};
+    sgx_qe_cert_data_t qe_cert_data = {0};
+
+    if (quote == NULL || pem_pck_certificate == NULL || pck_cert_chain == NULL)
+    {
+        OE_RAISE(OE_INVALID_PARAMETER);
+    }
+
+    OE_CHECK_MSG(
+        _parse_quote(
+            quote,
+            quote_size,
+            &sgx_quote,
+            &quote_auth_data,
+            &qe_auth_data,
+            &qe_cert_data),
+        "Failed to parse quote. %s",
+        oe_result_str(result));
+
+    *pem_pck_certificate = qe_cert_data.data;
+    *pem_pck_certificate_size = qe_cert_data.size;
+
+    // Read and validate the chain.
+    OE_CHECK(oe_cert_chain_read_pem(
+        pck_cert_chain, *pem_pck_certificate, *pem_pck_certificate_size));
+
+    result = OE_OK;
+done:
+
+    return result;
+}
+
+static void _update_validity(
+    oe_datetime_t* latest_from,
+    oe_datetime_t* earliest_until,
+    oe_datetime_t* from,
+    oe_datetime_t* until)
+{
+    if (oe_datetime_compare(from, latest_from) > 0)
+    {
+        *latest_from = *from;
+    }
+
+    if (oe_datetime_compare(until, earliest_until) < 0)
+    {
+        *earliest_until = *until;
+    }
+}
+
+oe_result_t oe_verify_quote_internal_with_collaterals(
+    const uint8_t* quote,
+    size_t quote_size,
+    const uint8_t* collaterals,
+    size_t collaterals_size,
+    oe_datetime_t* validation_time)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    uint8_t* local_collaterals = NULL;
+    size_t local_collaterals_size = 0;
+    oe_datetime_t validity_from = {0};
+    oe_datetime_t validity_until = {0};
+    oe_datetime_t creation_time = {0};
+
+    bool no_collaterals = false;
+
+    if (quote == NULL)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    if (collaterals == NULL)
+    {
+        result = oe_get_collaterals_internal(
+            quote,
+            quote_size,
+            (uint8_t**)&local_collaterals,
+            &local_collaterals_size);
+
+        if (result == OE_QUOTE_PROVIDER_CALL_ERROR)
+        {
+            // No qe_identity info returned from the quote provider, this could
+            // be because either get_qe_identity_info API was not supported or
+            // unexpected error. In both cases, check against hardcoded quoting
+            // enclave properties instead Assert that the qe report's MRSIGNER
+            // matches Intel's quoting. We will remove these hardcoded values
+            // once the libdcap_quoteprov.so was updated to support qe identity
+            // feature.
+            no_collaterals = true;
```

Comment:
I am not sure if above comments are accurate because there could be reasons other than qe_identity_id missing that could cause oe_get_collaterals_internal to fail if I read the code correctly

Reply:
The comment does mention unexpected errors. Let me know what would make it more clear.